htmd/smallmol/chemlab/builder.py

```diff
@@ -0,0 +1,406 @@
+
+
+class Builder:
+    """
+    A class to modify a SmallMol object: adding and removing hydrogens; creation of ligands from scracth
+
+    Parameters
+    ----------
+    smallmol: htmd.smallmol.smallmol.SmallMol
+        The SmallMol object
+        Default: None
+    checkInitialConformer: bool
+        If True, a conformation is generated if None are found. (designed espcially for smiles)
+        Default: True
+
+    Example
+    -------
+    >>> sm = SmallMol('CCC')
+    >>> sm.coords[0], sm.numAtoms
+   ( array([[ 0.],[ 0.],[ 0.]]), 11)
+   >>> B = Builder(sm)
+   >>> sm_new = B.getSmallMol()
+   >>> sm_new.coords[0]
+   array([[ 1.1701015 ],[-0.21571056],[-0.21615667]])
+   >>> B.removeHydrogens()
+   >>> B.getSmallMol().numAtoms
+   3
+    """
+
+    def __init__(self, smallmol=None, checkInitialConformer=True):
+
+        self.smallmol = smallmol
+        self.periodictable = PeriodicTable()
+
+        if checkInitialConformer and smallmol is not None:
+            self._prepareConformer()
+
+
+    def loadMol(self, smallmol, checkInitialConformer=True):
+        """
+        Loads the SmallMol object. The additional argument checkInitialConformer ensure that one valid conformer exists
+
+        Parameters
+        ---------
+        smallmol: htmd.smallmol.smallmol.SmallMol
+            The SmallMol object
+        checkInitialConformer: bool
+            If True, a conformer is generated if only not valid are found in the smallmol object
+            Default: True
+        """
+
+        if not isinstance(smallmol, SmallMol):
+            raise ValueError('The argument tyep {} is not accepted. Only SmallMol object. '.format(smallmol))
+
+        self.smallmol = smallmol.copy()
+
+        if checkInitialConformer:
+            self._prepareConformer()
+
+    def _prepareConformer(self):
+        """
+        Generate a conformer and stores the coords in the SmallMol object
+        """
+
+        from rdkit.Chem import AllChem
+
+        sm = self.smallmol
+        mol = sm.toRdkitMol(includeConformer=True)
+
+        AllChem.EmbedMolecule(mol, AllChem.ETKDG())
+
+        conf = mol.GetConformer(0)
+        coords = conf.GetPositions()
+        coords = coords[:, :, np.newaxis]
+        sm.coords = coords
+
+
+    def addHydrogens(self, onlyExplicit=False):
+        """
+        Adds the hydrogens to fill up the valence of the atoms. With onlyExplicit is possible to add only the polar ones
+
+        Parameters
+        ----------
+        onlyExplicit: bool
+            If True, only the polar hydrogens will be added
+            Default: False
+        """
+
+        pT = self.periodictable
+        sm = self.smallmol
+
+        # polar hydrogens are the ones bonded to hetero atoms (no carbon)
+        if onlyExplicit:
+            heavy_sel = ' '.join(_hetero_atoms)
+        else:
+            heavy_sel = ' '.join(_heavy_atoms)
+
+        atomidxs = sm.get( 'idx', 'element {}'.format(heavy_sel))
```

Comment:
If you end up calling this function many times, atomselections can become a bottleneck. I tend to work my way around them. For example, you could achieve the same selection with `np.in1d(sm.element, _heavy_atoms)` which returns a boolean mask with True if element is in _heavy_atoms

Reply:
Ah although this is the `SmallMol.get` so I guess you don't use atomselection. Ignore my comment. Sorry